pkg/controllers/ippool/ippool_controller.go

```diff
@@ -0,0 +1,249 @@
+}
+
+func (r *Reconciler) setReadyStatusFalse(ctx *context.Context, ippool *v1alpha2.IPPool, err *error) {
+	conditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionFalse,
+			Message: "NSX IPPool could not be created/updated",
+			Reason: fmt.Sprintf(
+				"error occurred while processing the IPPool CR. Error: %v",
+				*err,
+			),
+		},
+	}
+	ippool.Status.Conditions = conditions
+	if ippool.Status.Subnets == nil {
+		ippool.Status.Subnets = make([]v1alpha2.SubnetResult, 0)
+	}
+	e := r.Client.Status().Update(*ctx, ippool)
+	if e != nil {
+		log.Error(e, "unable to update IPPool status", "ippool", ippool)
+	}
+}
+
+func (r *Reconciler) setReadyStatusTrue(ctx *context.Context, ippool *v1alpha2.IPPool) {
+	conditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionTrue,
+			Message: "NSX IPPool has been successfully created/updated",
+			Reason:  "NSX API returned 200 response code for PATCH",
+		},
+	}
+	ippool.Status.Conditions = conditions
+	e := r.Client.Status().Update(*ctx, ippool)
+	if e != nil {
+		log.Error(e, "unable to update IPPool status", "ippool", ippool)
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &v1alpha2.IPPool{}
+	log.Info("reconciling ippool CR", "ippool", req.NamespacedName)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResType)
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch ippool CR", "req", req.NamespacedName)
+		return resultNormal, client.IgnoreNotFound(err)
+	}
+
+	// TODO: Since IPPool v1alpha2 service can only be activated from NSX 4.1.0 onwards,
+	// So need to check NSX version before starting IPPool reconcile
+	//if !r.Service.NSXClient.NSXCheckVersionForStaticRoute() {
+	//	err := errors.New("NSX version check failed, IPPool v1alpha2 is not supported")
+	//	updateFail(r, &ctx, obj, &err)
+	//	if NSX version check fails, it will be put back to reconcile queue and be reconciled after 5 minutes
+	//return ResultRequeueAfter5mins, nil
+	//}
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResType)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.IPPoolFinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.IPPoolFinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "ippool", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return resultRequeue, err
+			}
+			log.V(1).Info("added finalizer on ippool CR", "ippool", req.NamespacedName)
+		}
+
+		subnetCidrUpdated, ipPoolSubnetsUpdated, err := r.Service.CreateOrUpdateIPPool(obj)
+		if err != nil {
+			log.Error(err, "operate failed, would retry exponentially", "ippool", req.NamespacedName)
```

Comment:
if this infinite retry? If in some case, create IP Pool failed due to nsx manager down or some other un-recoverable error, should we also retry infinitely?

Reply:
Yes, all the CRDs would retry infinitely since the event exists in the queue forever.

Reply:
I mean for some cases, if retry won't work, then we do not need to requeue the IPPool CR.

Reply:
I think it is a common feature for all CRDs, should we discuss it and classify the cases for all CRDs when we must requeue or not?

Reply:
TODO added